go.go

```diff
@@ -99,35 +99,55 @@ func Go(a, b []byte) (changes []*Change, err error) {
 			ds := x.diffstat(ai, bi)
 			changes = append(changes, &Change{Name: s, DelLines: ds.del, InsLines: ds.ins})
 		} else {
-			changes = append(changes, &Change{Name: s, Inserted: true, InsLines: len(bytes.Split(x.bBytes(bi), newline))})
+			changes = append(changes, &Change{Name: s, Inserted: true, InsLines: x.bLines(bi)})
 		}
 	}
 	for s, ai := range del {
 		if _, ok := ins[s]; !ok {
-			changes = append(changes, &Change{Name: s, Deleted: true, DelLines: len(bytes.Split(x.aBytes(ai), newline))})
+			changes = append(changes, &Change{Name: s, Deleted: true, DelLines: x.aLines(ai)})
 		}
 	}
 
 	if other != nil {
-		// Calculate an overall diffstat.
-		aLines := bytes.Split(a, newline)
-		bLines := bytes.Split(b, newline)
-		ab := diff.Bytes(aLines, bLines)
-		es := diff.Myers(context.Background(), ab)
-		ins, del := es.Stat()
-		// Subtract all diffs accountable for by other changes.
-		for _, c := range changes {
-			ins -= c.InsLines
-			del -= c.DelLines
+		// Calculate diffstats between non-declarations.
+		aIsLarger := false
+		lens := len(x.asplit) - 1
+		if len(x.asplit) > len(x.bsplit) {
+			lens = len(x.bsplit) - 1
+			aIsLarger = true
 		}
-		if ins < 0 {
-			ins = 0
+		// If df is increased, it means that an index has been displaced by some change.
+		var df int64
+		for i := 1; i < lens; i++ {
+			if int64(x.asplit[i]) != (int64(x.bsplit[i]) - df) {
+				df = abs(int64(x.asplit[i] - x.bsplit[i]))
+				// "other" changes may happen at even indices only.
```

Comment:
I'm a bit nervous about hard-coding this assumption. For example, if we wanted to expand to include other top level decls (`type`, `var`, `import`, `const`), it's not obvious to me that the even/odd pattern is one we would want to keep.

Reply:
I was worried about this too, that's why I included tests with inlined functions. However, the only thing that appears to matter is that 2  indices are inserted per declaration, just as is being done now when constructing ```x.asplit``` and ```x.bsplit``` at the beginning:
```
for ... {
  ...
  x.asplit = append(x.asplit, afset.Position(d.Pos()).Offset)
  ...
  x.asplit = append(x.asplit, afset.Position(d.End()).Offset)
}
```
I'm sure it holds with func decls, otherwise the tests would have failed already. I can't guarantee it will hold with other types of decls (if we keep adding 2 indices per decl, it should), so maybe we could add an ```if ... { panic() }``` for now, so that we can debug this easily if it breaks when adding other decls?